R/utilities-simulation-results.R

```diff
@@ -223,59 +223,30 @@ simulationResultsToDataFrame <- function(simulationResults,
     values_to = "simulationValues"
   )
 
-  # Extract units and dimensions for paths in a separate data frame
-  # iterate over the list (in `.x`) using index (names for list elements)
-  # and apply function in `.f` to each element.
-  #
-  # The result will be a list of data frames, which will be row-wise glued into
-  # a single data frame with the `_dfr` variant of this function.
-  simMetaData <- purrr::imap_dfr(
-    .x = simList$metaData,
-    .f = ~ as.data.frame(.x, row.names = NULL, stringsAsFactors = FALSE),
-    .id = "paths"
-  )
-
-  # Leave out time units and dimensions since time is not a path; they will be
-  # added at a later stage.
-  simMetaData <- dplyr::filter(simMetaData, paths != "Time")
-
-  # Combine data frame with simulated data and meta data.
-  simData <- dplyr::left_join(simData, simMetaData, by = "paths")
-
-  # Add back in the previously left out time meta data to the combined data frame.
-  simData <- dplyr::bind_cols(
-    simData,
-    dplyr::tibble(
-      "TimeUnit" = simList$metaData$Time$unit[[1]],
-      "TimeDimension" = simList$metaData$Time$dimension[[1]]
+  units <- lapply(simList$metaData, function(path) {
+    path$unit
+  })
+  dims <- lapply(simList$metaData, function(path) {
+    path$dimension
+  })
+  molWeights <- lapply(unique(simData$paths), function(path) {
+    molWeight <- ospsuite::toUnit(
+      quantityOrDimension = ospDimensions$`Molecular weight`,
+      values              = simulationResults$simulation$molWeightFor(path),
+      targetUnit          = ospUnits$`Molecular weight`$`g/mol`
     )
-  )
+  })
+  names(molWeights) <- unique(simData$paths)
 
-  # For each path, extract the molecular weight based on that path string
-  #
-  # This involves first grouping and nesting the data by path. Note that
-  # `nest()` here will have a better performance than `rowwise()`. E.g., if
-  # there are 100 rows, `rowwise()` will run the computation 100 times, while
-  # with `nest()`, the computation only be carried for the same number of
-  # times as the number of `paths` present.
-  simData <- simData %>%
-    dplyr::group_by(paths) %>%
-    tidyr::nest() %>%
-    # Add a new column for molecular weight.
-    # When you call `molWeightFor()`, it returns the value in the base unit -
-    # which is `kg/µmol`. This is not the unit the user would expect, so we
-    # convert it first to the common unit `g/mol`.
+  simData <- dplyr::group_by(simData, paths) %>%
```

Comment:
@Felixmil Also here - any speed up ideas?

Reply:
this tibble operation should not be very long, however we can still improve it a bit for this kind of column mutation using `rowwise`:
```r
  simData <-
    simData %>%
    dplyr::rowwise() %>%
    dplyr::mutate(
      TimeDimension = dims$Time,
      TimeUnit = units$Time,
      dimension = dims[[paths]],
      unit = units[[paths]],
      molWeight = molWeights[[paths]]
    )
```

Reply:
This actually DOUBLED the time needed :D

Reply:
Here is the solution using `{data.table}`:

```r
  simData <-  data.table::as.data.table(simData)
  simData <- simData[,`:=`(TimeDimension =  dims$Time,
                           TimeUnit =  units$Time,
                           dimension =  dims[[paths]],
                           unit =  units[[paths]],
                           molWeight = molWeights[[paths]]),
                     by = paths]
```
Result: ~2 times faster. This should be even faster with large datasets
```
Unit: milliseconds
                                     expr    min      lq     mean median      uq     max neval
     og(simData, dims, units, molWeights) 4.3466 4.89990 5.936743 5.0822 5.40190 22.0154   100
    new(simData, dims, units, molWeights) 3.9706 4.34070 5.003569 4.5498 4.91425 18.0941   100
 new_dt(simData, dims, units, molWeights) 1.4282 2.07735 3.029173 2.3516 2.65680 16.4108   100
```

Reply:
Results with a simData with `2 946 000` rows:
```r
Unit: milliseconds
                                         expr      min       lq     mean   median       uq      max neval
     og(big_simData, dims, units, molWeights) 204.5205 219.9517 326.0187 322.4694 430.7714 449.7938    10
    new(big_simData, dims, units, molWeights) 157.8323 180.3917 232.1167 187.7086 320.0275 374.1119    10
 new_dt(big_simData, dims, units, molWeights) 101.9321 119.9672 180.7313 123.3568 303.0376 340.7760    10
```